Encrypted logs

oguzkocer opened the pull request:
This PR introduces encrypted logging support. It closely follows the FluxC conventions, but there are some minor differences due to this being a separate functionality from WordPress or Woo.

The intention for this design is for the client to ask FluxC to upload a given `File` and forget about it. There is some internal queuing logic which utilizes the DB for state keeping. The encryption logic has been brought in from the WPAndroid, thanks to @jkmassel's implementation and only minor modifications have been made to make it fit the FluxC ecosystem.

The best way to test and review this PR might be to run the connected tests in `ReleaseStack_EncryptedLogTest` and follow the logic with breakpoints. One gotcha to keep in mind is the `too_many_requests` error which I was hitting fairly quickly. The current tests are designed to ignore this error since there is nothing we can do about them on our end. We were either going to ignore them or not have these tests at all, since we wouldn't want to deal with randomly failing tests if we couldn't fix them. We may add some tests where we mock the network layer later on, but that'd provide very low value right now.

I left some todos in the code for future improvements and targeted a feature branch for this PR to give us a chance to work on them. The only thing that I couldn't figure out yet is bringing in [EncryptionUtilsTest](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/androidTest/java/org/wordpress/android/util/EncryptionUtilsTest.kt) from WPAndroid. I've spent way too much time on this and would love to get some help on the build issues it's causing. @jkmassel Do you think they are high value tests or should we just not add them if we can't get them to work quickly?

@jkmassel I assume you'll be the main reviewer for this PR. Please let me know if you want to go over these changes together.

@loremattei You may want to read this description and take a quick look at the few todos I left in the code to get a sense of where we are and whether any of it is something we should prioritize.

jkmassel commented on fluxc/src/main/java/org/wordpress/android/fluxc/store/EncryptedLogStore.kt:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    private fun deleteEncryptedLog(encryptedLog: EncryptedLog) {
+        // TODO: Do we want to delete the unencrypted log file?
```
I would say no – we might as well leave the log file around until it's cleaned up – there are other useful debugging things associated with this (ie – allowing a user to view their own logs)

oguzkocer replied:
That's what I thought as well, but just to make sure we discuss this, I left this todo.

oguzkocer replied:
Removed in ae728cea808b57138ba296252bd9ab69a66880ad.

jkmassel commented on fluxc/src/main/java/org/wordpress/android/fluxc/store/EncryptedLogStore.kt:
```diff
@@ -0,0 +1,183 @@
+        }
+    }
+
+    // TODO: Any other known upload errors we should handle?
```
Anything in the 500 range should IMHO be treated like a 429 Rate Limit error – just try again later, but don't count it as a failed retry. WP.com doesn't provide the `Retry-After` header, which is why I used exponential backoff on iOS instead.

oguzkocer replied:
Improved in 33c052fe659aee6387154bf1c1f319186927b260. We should add any other known errors as specific cases if/when they come up. However, with this change, the current error handling should be good enough for the first version 🤞

jkmassel commented on fluxc/src/main/java/org/wordpress/android/fluxc/store/VerticalStore.kt:
```diff
@@ -33,7 +33,7 @@ class VerticalStore @Inject constructor(
 
     override fun onRegister() {
-        AppLog.d(AppLog.T.API, ListStore::class.java.simpleName + " onRegister")
+        AppLog.d(AppLog.T.API, VerticalStore::class.java.simpleName + " onRegister")
```
Heh nice fix

oguzkocer commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/encryptedlog/EncryptedLogRestClient.kt:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    // {"error":"too_many_requests","message":"You're sending too many messages. Please slow down."}
+    // {"error":"invalid-request","message":"Invalid UUID: uuids must only contain letters, numbers, dashes, and curly brackets"}
```
If it's an easy change, that'd be great.

I also didn't mean to leave these jsons as comments, but should we maybe keep them? It may be kind of nice to be able to see the exact error json without debugging. 🤔

oguzkocer commented on example/src/test/java/org/wordpress/android/fluxc/encryptedlog/EncryptedLogSqlUtilsTest.kt:
```diff
@@ -0,0 +1,152 @@
+        dateCreated: Date = Date(),
+        uploadState: EncryptedLogUploadState = QUEUED
+    ) = EncryptedLog(
+            // Bypass the annoying milliseconds comparison issue
```
Yes. It was originally a comment on the property creation, but I used AS to convert it to return the method with `=` which has put the comment there. I just didn't notice it. Thanks for catching it!

oguzkocer replied:
Fixed in ae728cea808b57138ba296252bd9ab69a66880ad.

oguzkocer commented:
@jkmassel @loremattei Thank you for the review. I've addressed the comments and fixed the merge conflict. It's ready for a second pass.

> I do think we should include the EncryptionUtilsTest, just in case we're messing with the implementation in the future. But no reason that couldn't be a separate PR. I could take a run at it as well if you'd like – I apparently got it working once before? 😜

@jkmassel If you have some extra time, I'd really appreciate if you could take a look. I bet it's something super simple, but I can't seem to figure out the proper setup for it. I think the issue might be that we are running our tests from a different project, so we have to add the dependency to the `example` project as well to fix the import issue. However, even after doing that there is a class-linking error (I think that's what it was saying) that I can't figure out. I tried adding the tests as both unit tests and instrumentation tests with no luck. If you can get `keypair = EncryptionUtils.sodium.cryptoBoxKeypair()` line to work for a simple test, I can bring in the rest 🤞 

> I think this is definitely ready to move forward, though once we have it wired into the actual app I think I'd step through it all again in a non-test context to test thoroughly that way as well.

Sounds great!

oguzkocer commented:
@jkmassel Since all the comments are addressed, the PR is approved and this is targeting a feature branch, I am going to merge this PR to simplify the process. (will open a new PR in a bit) However, please let me know if there is anything I missed from the review and I'll address it as a separate PR.

Also, please disregard my call for help for `EncryptionUtilsTest` as I was able to get that working and will include in the new PR.
